### DISCUSSION TITLE
Unnecessary import of `nodemailer` for Email Provider

### DISCUSSION THREAD

**Discussion by anguskeatinge** [nextauthjs/next-auth#3172] [Help] [answered]
### Description 🐜

error - Error: Cannot find module 'nodemailer'

I have a custom sign in page and a custom sendVerificationRequest so nodemailer is a superfluous dependency.

### Is this a bug in your own project?

No

### How to reproduce ☕️

Don't install peer dependency nodemailer when using email provider.

### Environment 🖥

Not necessary

### Contributing 🙌🏽

Yes,  I am willing to help solve this bug in a PR

**Comment by anguskeatinge** [comment]
Here is what I was thinking:
https://github.com/nextauthjs/next-auth/pull/3168

**Comment by iaincollins** [comment]
I think @balazsorban44 had some good ideas about this and is either already doing this in v4 or is planned for the future.

**Comment by balazsorban44** [comment, marked as answer]
So this was not a bug, we already separated `nodemailer` from the core code. In case you don't use the Email provider, you won't need to install it. It is mentioned in the migration guide that you have to install it if you plan to use the built-in Email provider https://next-auth.js.org/getting-started/upgrade-v4#nodemailer

I have tried different ways to conditionally require `nodemailer`, but neither `require` nor dynamic import wasn't future proof/sufficient.

For now, if you really don't want to install `nodemailer`, just take this file and create your own email provider: https://github.com/nextauthjs/next-auth/blob/beta/src/providers/email.ts

Most of the settings are `nodemailer` specific anyway!

In the future, we could maybe do something like with the Adapters, where we expect the user to give us a client/instance, so it always matches their already-installed version.

**Reply by iaincollins** [reply]
Ah thanks! I remember you going into detail about that somewhere now you say that again.

**Reply by anguskeatinge** [reply]
Yep - makes a lot of sense. Thanks @balazsorban44 .

Tbh it's hardly any skin off my shoulders to include a dependency that I don't use. I should clean up my lodash imports before complaining about this 🤣 

Interesting what you say about it not being tree shaken: https://github.com/nextauthjs/next-auth/pull/3168#issuecomment-965821531

**Comment by IuliiaBondarieva** [comment]
I had this error after I tried to override JWT type with
```
declare module '@auth/core/jwt'
```
after changing it to 
```
declare module 'next-auth/jwt' {
```
error disappeared.
